### DISCUSSION TITLE
Operator 0.31.0 is not scaling to the requested number of replicas

### DISCUSSION THREAD

**Discussion by popdaniel942** [strimzi/strimzi-kafka-operator#7341] [Q&A] [answered]
Hello. I am using the kafka operator (0.31.0) and I have a cluster with 3 replicas. This creates 3 kafka pods and 3 zookeeper pods with statefulsets (I am using the featuregate to not use StrimziPodSets). If I scale 1 statefulset to one pod, the operator will put it back up to 3 pods. But if I scale both statefulsets to 1 pod, the operator is not puting them back up to 3 pods. It fails with timeout exceeded and zookeeper errors because there are not 3 pods running. With strimzi operator 0.27.1 this problem is not happening for me. Am I doing something wrong? Is it supposed to not scale back up the statefulsets?

**Comment by scholzj** [comment, marked as answer]
You should never scale the StatefulSets. If you want to change the number of replicas, you should always do it in the `Kafka` CR.

**Reply by popdaniel942** [reply]
I agree. It's just that I didn't have this problem with an older version of the operator. I was scaling all statefulsets and deployments to 0, inluding the strimzi operator, and then I was scaling everything back to 1. And the operator knew to put 3 pods back up in the kafka statefulsets. Also this way, the pvcs would not get deleted. If I change the number of replicas in the CR, wouldn't some PVCs get deleted?

**Reply by scholzj** [reply]
Well, I suspect the manual scaling of the ZooKeeper cluster broke it. If you want to manipulate the StatefulSets, you always need to switch of the operator and pause the reconciliation first. You also need to keep in mind that the operator configures the pods. So after you disable the operator, you can in theory delete all the StatefulSets / Pods (or scale them to 0) to _stop_ them for the weekend or something like that. But you cannot scale them up or down as you want. That requires other configuration changes as well.

**Reply by popdaniel942** [reply]
well it is not letting me to put the number of replicas to 0 in the Kafka CRD. I can put it on any number but not 0. And I need to scale everything down to 0.
This is the config used for 3 replicas for kafka and zookeeper in the CRD.
config:
      auto.create.topics.enable: false
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      default.replication.factor: 1
      min.insync.replicas: 1
      inter.broker.protocol.version: "3.2"

**Reply by scholzj** [reply]
We never supported scaling to 0 as it doesn't make sense for a stateful application. That definitely didn't changed since previous versions.

**Reply by popdaniel942** [reply]
but in the previous version, in our case 0.27.1, we were able to scale down the statefulsets from 3 to 0 and the operator deployment from 1 to 0, and then we scaled the operator back up to 1, it knew to put 3 pods in each statefulset. Now this workaround does not work anymore, the operator does not know how to put the 3 pods back up, probably something in the backend fails. We were using this to stop all pods during the night and in the weekends, but now it's not working anymore.

**Reply by scholzj** [reply]
Sorry, but I'm very confused what are you actually asking. You started talking about scaling to from 3 to 1 and said what does not work there for you. Now you talk about scaling to 0, which is something different and you never explained the issue there. So maybe you should describe what exactly are you doing step by step.

**Reply by popdaniel942** [reply]
It doesn't really matter if I scale the statefulsets to 0 or 1. If I scale both stateful down or up, the latest operator (also 0.30.0) doesn't reput the number of replicas said in the kafka CR. I understand that for 1 replica we can change the number in the kafka CR. For 0 replicas I guess the best solution is to remove the CR completely. I thought the PVCs will be deleted but I saw now that there is a flag to not delete them. So I have the solution now for my problem. But I still believe there were some changes on the operator which made it to not support scaling down or up the statefulsets. Because even if I stop the operator and then I scale the statefulsets and then I reput the operator, the reconciliation does not seem to work, while with an older version of the operator it works.

**Reply by scholzj** [reply]
It **does** matter what you do and in which order you do it. But when you don't explain properly what you do, it is hard to understand what exactly is happening ... understanding what caused it is even harder.

There might have been changes to the operator, of course - that is at the end the main reason why we do new releases. And since scaling the StatefulSet is something you are not supposed to do, there is obviously no guarantee that what worked a year ago will still work today.

**Reply by popdaniel942** [reply]
You are right. I appreciate your answers. Now I understand better what needs to be done in my case. Sorry for not explaining properly. Thank you.
